libs/muddle/src/peer_selector.cpp

```diff
@@ -403,36 +414,71 @@ void PeerSelector::OnAnnouncement(Address const &from, byte_array::ConstByteArra
     }
   }
 
-  // update the peers info cache
+  // remove all internal references to this objects address and peer
+  std::unordered_set<Address> removed_addresses{};
   {
-    auto it = peers_info_.find(from);
-    if (it != peers_info_.end())
+    auto it = peers_info_.begin();
+    while (it != peers_info_.end())
     {
-      Metadata &metadata       = it->second;
-      PeerData &existing_peers = metadata.peer_data;
+      Metadata &metadata  = it->second;
+      auto &    peer_data = metadata.peer_data;
 
-      for (auto const &peer : peers)
+      // remove the peer address from the cache
+      bool peers_removed{false};
+      for (auto peer_it = peer_data.begin(); peer_it != peer_data.end();)
       {
-        // attempt to locate the address in the peer list
-        bool const peer_present = std::find_if(existing_peers.begin(), existing_peers.end(),
-                                               [&peer](PeerMetadata const &entry) {
-                                                 return entry.peer == peer;
-                                               }) != existing_peers.end();
-
-        if (!peer_present)
+        if (core::IsIn(peers, peer_it->peer))
         {
-          metadata.peer_data.emplace_back(PeerMetadata{peer});
+          peer_it       = peer_data.erase(peer_it);
+          peers_removed = true;
+        }
+        else
+        {
+          ++peer_it;
         }
       }
-    }
-    else  // new entry
-    {
-      Metadata &metadata = peers_info_[from];
 
-      for (auto const &peer : peers)
+      // remove the entry completely if we have removed all previous peers
+      if (peers_removed && peer_data.empty())
       {
-        metadata.peer_data.emplace_back(PeerMetadata{peer});
+        removed_addresses.emplace(it->first);
+        it = peers_info_.erase(it);
       }
+      else
+      {
+        ++it;
+      }
+    }
+  }
+
+  // for all the addresses that have been removed, they must be removed from the desired and kad
+  // peers lists too.
+  desired_addresses_  = desired_addresses_ - removed_addresses;
+  kademlia_addresses_ = kademlia_addresses_ - removed_addresses;
+
+  Metadata *metadata{nullptr};
+
+  // lookup or create the metadata object
+  auto it = peers_info_.find(from);
+  if (it != peers_info_.end())
+  {
+    // lookup existing entry
+    metadata = &(it->second);
+  }
+  else
+  {
+    // create a new entry
+    metadata = &peers_info_[from];
+  }
+
+  if (metadata != nullptr)
```

Comment:
Can't be false.

Reply:
true with the code as it is at the moment, this is just a belt and braces check for the future 😁